### PR TITLE
Curvature drive documentation

### PR DESCRIPTION
**Pull request by hyperliskdev**

### CODE REVIEW

**Comment by Ewpratten** [comment]
You appear to have upgraded the website to Jekyll 4.0

May need to revert this, as we should be running 3.x

**Comment by Ewpratten** [comment]
Tons of spelling errors.. Ill try to fix most

**Comment by Ewpratten** [comment]
Also, there is a lot of un-explained code. This may need some big changes

**Comment by Ewpratten** [comment]
And `/curv` is not a great permalink in any way

**Comment by hyperliskdev** [comment]
Well, thats not good. I am fixing spelling and making things a little more specific. Where would the site have been updated? The Gemfile I have says its 3.8.6?

**Comment by Ewpratten** [comment]
one sec.. writing a proper request for changes..

File: docs/learn/curvature-drive.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+Curvature drive is a method inside of the DifferentialDrive class in WPILib 
+
+## What does Curvature Drive do?
+
+Curvature drive is a type of drive just like Arcade or Tank drive. Curvature drive is different in a few key ways;
+
+- isQuickTurn method is used to quickturn on the stop, as it is default off.
+- Turning speed withput isQuickTurn == true, is based on how fast the robot is moving.
+- The robot can be driven with either Trigger Control or Joystick Control
+- And finally, the robot now drives like a car.
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: None of these make any sense.. at all. Here is some info you could turn into points

 - QuickTurn is used to switch to rate-of-change control, which allows on-the-spot turning
 - rate-of-change control provides tank-like control
 - constant-curvature control feels more like a car

File: docs/learn/curvature-drive.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+    m_DifferentialDrive.curvatureDrive(speed, rotation, isQuickTurn);
+    
+    
+  }
+```
+
+Is moving isnt required but could be helpful in auto or for something else ive never heard of.
+
+Inside the drive controller class:
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: There should be more words here.. Probably explain how to pass inputs to the drivetrain method?

File: docs/learn/curvature-drive.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+``` java
+// Called repeatedly when this Command is scheduled to run
+  @Override
+  protected void execute() {
+
+
+    // Main vars for driving.
+    speed = Robot.m_oi.getTriggerThrottle();
+    rotation = Robot.m_oi.getRotation();
+    isQuickTurn = Robot.m_oi.getQuickTurn();
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: These are all specific to HATCHfield and MiniBot.. Not useful anywhere else.. Remove them, or provide commented stubs

File: docs/learn/curvature-drive.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+    // Deadbanding stuff
+    speed = m_speedDeadband.feed(speed);
+    rotation = m_rotationDeadband.feed(rotation);
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: These have nothing to do with drive.. Did you just copy-paste our codebase into a MD file?

File: docs/learn/curvature-drive.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    Robot.m_driveTrain.curvDrive(speed, rotation, isQuickTurn);
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: This should be more general.. maybe `DriveTrain.curvatureDrive(.....)`?

File: docs/learn/curvature-drive.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+
+  }
+
+```
+
+getTrigger will set the speed to the 0 --> 1 vaule it outputs <br>
+getRotation get the x axis of the right joystick values from -1 --> 0 --> 1 <br>
+quickTurn will let the Robot turn without speed. <br>
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: I get what you are trying to say.. But it could be worded much better

File: docs/learn/curvature-drive.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+If quickTurn is not enabled, the robot will not turn unless it is moving.
+When the robot is moving, it will only turn based on how fast it is moving.
+Spesifically, this fast
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: What are these words?

File: docs/learn/curvature-drive.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+``` java
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: You should describe what this is.. Maybe title this section "How constant-curvature control works"

File: docs/learn/curvature-drive.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+      angularPower = Math.abs(xSpeed) * zRotation - m_quickStopAccumulator;
+
+      if (m_quickStopAccumulator > 1) {
+        m_quickStopAccumulator -= 1;
+      } else if (m_quickStopAccumulator < -1) {
+        m_quickStopAccumulator += 1;
+      } else {
+        m_quickStopAccumulator = 0.0;
+      }
+    }
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: Not really how constant-curvature control works.. This is all WPILib-specific stuff, and references code that you didnt share..

Here is the actual code that makes it work
```java

if(quickTurn){
 L = speed + rotation;
 R = speed - rotation;
} else {
 L = speed * abs(speed) + rotation;
 R = speed * abs(speed) - rotation;
}
```

File: docs/learn/curvature-drive.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+```
+angularPower is a double, that; if IsQuickTurn is true, will be default set to the rotation given. <br>
+if not; the angularPower is set to the absolute value of speed * rotation - this weird quickStopAccumulator. <br>
+The quickstop accumulator isnt actually, initalized inside of DifferrentialDrive so i have no clue what that value is doing... <br>
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: All WPILib-specific stuff that doesn't matter..

File: docs/learn/curvature-drive.md [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+- isQuickTurn method is used to quickturn on the stop, as it is default off.
+- Turning speed withput isQuickTurn == true, is based on how fast the robot is moving.
+- The robot can be driven with either Trigger Control or Joystick Control
+- And finally, the robot now drives like a car.
+
+## How to make Curvature drive
+
+```java
+
+// Inside DriveTrain.java (hyperliskdev/HyperBot)
```
Comment (Ewpratten) [inline docs/learn/curvature-drive.md]: Dont link to your own projects.. This should be codebase-independant. Also, who cares what repo its from?

Comment (hyperliskdev) [reply docs/learn/curvature-drive.md]: Sorry, referencing things is something I thought was important. That was just a citation.

File: docs/learn/curvature-drive.md [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+- isQuickTurn method is used to quickturn on the stop, as it is default off.
+- Turning speed withput isQuickTurn == true, is based on how fast the robot is moving.
+- The robot can be driven with either Trigger Control or Joystick Control
+- And finally, the robot now drives like a car.
+
+## How to make Curvature drive
+
+```java
+
+# Inside of my drivetrain class.
```
[the remark this answers is not in the document]
Comment (hyperliskdev) [reply docs/learn/curvature-drive.md]: Yep...

**Comment by Ewpratten** [comment]
#7

**Comment by Ewpratten** [comment]
> Well, thats not good. I am fixing spelling and making things a little more specific. Where would the site have been updated? The Gemfile I have says its 3.8.6?

odd.. Ill check that

**Comment by hyperliskdev** [comment]
`source "https://rubygems.org"

# Hello! This is where you manage which Jekyll version is used to run.
# When you want to use a different version, change it below, save the
# file and run `bundle install`. Run Jekyll with `bundle exec`, like so:
#
#     bundle exec jekyll serve
#
# This will help ensure the proper Jekyll version is running.
# Happy Jekylling!
gem "jekyll", "~> 3.8.5"

# Docs theme
gem "just-the-docs"

# If you want to use GitHub Pages, remove the "gem "jekyll"" above and
# uncomment the line below. To upgrade, run `bundle update github-pages`.
# gem "github-pages", group: :jekyll_plugins

# If you have any plugins, put them here!
group :jekyll_plugins do
  gem "jekyll-feed", "~> 0.6"
end

# Windows does not include zoneinfo files, so bundle the tzinfo-data gem
gem "tzinfo-data", platforms: [:mingw, :mswin, :x64_mingw, :jruby]

# Performance-booster for watching directories on Windows
gem "wdm", "~> 0.1.0" if Gem.win_platform?

gem 'jekyll-admin', group: :jekyll_plugins
gem 'jekyll-redirect-from'
gem 'jemoji'
gem 'jekyll-mentions'


# gem "nokogiri", "~> 1.7.0"

`

I'm not sure why you would have had to install 48 new gems. This is my Gemfile right now.

**Comment by Ewpratten** [comment]
Now that I think about it, the whole Jekyll version thing might be caused by us having different plugins downloaded on our computers? Let's ignore that issue, and Ill add the `Gemfile.lock` to the `.gitignore` after merge.

**Comment by Ewpratten** [comment]
#24 contains a fix for the Gemfile, and a fast-forward to master

**Comment by Ewpratten** [comment]
#24 has been merged. Make sure to pull from origin

**Pull request merged**